sympy/tensor/tensor.py

```diff
@@ -2632,9 +2632,12 @@ def canon_bp(self):
         under monoterm symmetries.
         """
         expr = self.expand()
-        args = [canon_bp(x) for x in expr.args]
-        res = TensAdd(*args).doit(deep=False)
-        return res
+        if isinstance(expr, self.func):
+            args = [canon_bp(x) for x in expr.args]
+            res = TensAdd(*args).doit(deep=False)
+            return res
+        else:
+            return canon_bp(expr)
```

Comment:
If you have a case that doesn't expand to be a self.func type, will expanding again succeed? Guessing, not, I suggest just returning the expr instead of recursing.